### PR TITLE
[W5.4][F10-A4]Benjamin Ang

### PR DESCRIPTION
**Pull request by NUSe0032202**
Added a sort command to sort the address book by name alphabetically, code is ready for review.

### CODE REVIEW

File: src/seedu/addressbook/parser/Parser.java [hunk 1]
```diff
@@ -11,17 +11,7 @@
...
-import seedu.addressbook.commands.Command;
-import seedu.addressbook.commands.DeleteCommand;
-import seedu.addressbook.commands.ExitCommand;
-import seedu.addressbook.commands.FindCommand;
-import seedu.addressbook.commands.HelpCommand;
-import seedu.addressbook.commands.IncorrectCommand;
-import seedu.addressbook.commands.ListCommand;
-import seedu.addressbook.commands.ViewAllCommand;
-import seedu.addressbook.commands.ViewCommand;
+import seedu.addressbook.commands.*;
```
Comment (aver0214) [inline src/seedu/addressbook/parser/Parser.java]: Chann to import seedu.addressbook.commands.SortCommand?

File: src/seedu/addressbook/commands/SortCommand.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    public static final String COMMAND_WORD = "sort";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Sorts the entries in the address book by name alphabetically.\n"
+            + "Example: " + COMMAND_WORD;
+    public static final String MESSAGE_SUCCESS = "The address book has been sorted alphabetically according to name.";
+
+    public SortCommand(){}
+
+    @Override
+    public int compare(Person a, Person b){
```
Comment (aver0214) [inline src/seedu/addressbook/commands/SortCommand.java]: When is this compare method executed?

File: src/seedu/addressbook/parser/Parser.java (hunk 1 above)
Comment (wenmogu) [reply src/seedu/addressbook/parser/Parser.java]: the star is to import everything under seedu.addressbook.commands so i guess its ok (if i understand ur qn correctly...)

File: test/input.txt [hunk 3]
```diff
@@ -149,6 +149,20 @@
   list
 
 ##########################################################
+# test sort command
```
Comment (wenmogu) [inline test/input.txt]: Hello Ben! 
Good job! i like how u did things incrementally and the logic is clear!
Emm probably for this I/O test, you might also want to test the occasion when it sorts an empty list, just to see if the message shown is correct:)

File: src/seedu/addressbook/parser/Parser.java (hunk 1 above)
Comment (whipermr5) [reply src/seedu/addressbook/parser/Parser.java]: Avoid star imports as this violates the [coding standard](https://oss-generic.github.io/process/codingStandards/CodingStandard-Java.html#package-and-import-statements).

File: src/seedu/addressbook/commands/SortCommand.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        return a.getName().fullName.compareToIgnoreCase(b.getName().fullName);
+    }
+
+    @Override
+    public CommandResult execute() {
+        try {
+            addressBook.sort();
+            return new CommandResult(MESSAGE_SUCCESS);
+        }catch(ListIsEmptyException lie){
```
Comment (whipermr5) [inline src/seedu/addressbook/commands/SortCommand.java]: Missing spaces.

File: src/seedu/addressbook/commands/SortCommand.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Sorts the entries in the address book by name alphabetically.\n"
+            + "Example: " + COMMAND_WORD;
+    public static final String MESSAGE_SUCCESS = "The address book has been sorted alphabetically according to name.";
+    public static final String MESSAGE_LIST_EMPTY = "The address book is currently empty, please add at least two entries.";
+
+    public SortCommand(){}
+
+    @Override
+    public int compare(Person a, Person b){
```
Comment (whipermr5) [inline src/seedu/addressbook/commands/SortCommand.java]: Missing spaces.

File: src/seedu/addressbook/data/AddressBook.java [hunk 6]
```diff
@@ -128,4 +129,7 @@ public boolean equals(Object other) {
                         && this.allPersons.equals(((AddressBook) other).allPersons)
                         && this.allTags.equals(((AddressBook) other).allTags));
     }
+    public void sort()throws ListIsEmptyException{
```
Comment (whipermr5) [inline src/seedu/addressbook/data/AddressBook.java]: Missing spaces and newline before.

File: src/seedu/addressbook/data/person/UniquePersonList.java [hunk 7]
```diff
@@ -28,6 +29,9 @@ protected DuplicatePersonException() {
             super("Operation would result in duplicate persons");
         }
     }
+    public static class ListIsEmptyException extends Exception{
+        protected ListIsEmptyException() {super("Operation would not work as the address book is empty");}
```
Comment (whipermr5) [inline src/seedu/addressbook/data/person/UniquePersonList.java]: Same here.

File: src/seedu/addressbook/data/person/UniquePersonList.java [hunk 8]
```diff
@@ -28,6 +29,9 @@ protected DuplicatePersonException() {
             super("Operation would result in duplicate persons");
         }
     }
+    public static class ListIsEmptyException extends Exception{
```
Comment (whipermr5) [inline src/seedu/addressbook/data/person/UniquePersonList.java]: Same here.

File: src/seedu/addressbook/data/person/UniquePersonList.java [hunk 9]
```diff
@@ -140,4 +144,11 @@ public boolean equals(Object other) {
                 || (other instanceof UniquePersonList // instanceof handles nulls
                         && this.internalList.equals(((UniquePersonList) other).internalList));
     }
+    public void sort() throws ListIsEmptyException{
+       if(!internalList.isEmpty()) {
+           Collections.sort(internalList, new SortCommand());
+       }else {
```
Comment (whipermr5) [inline src/seedu/addressbook/data/person/UniquePersonList.java]: Same here.

File: src/seedu/addressbook/commands/SortCommand.java (hunk 2 above)
Comment (whipermr5) [reply src/seedu/addressbook/commands/SortCommand.java]: This will become clearer when interfaces are covered this week.
